### DISCUSSION TITLE
Install zha-quirks on new setup

### DISCUSSION THREAD

**Discussion by BashSer** [zigpy/zha-device-handlers#1595] [Q&A]
Good day !
I just can't figure out how to install it.
I write in configuration.yaml this:
```
zha:
  enable_quirks: true
  custom_quirks_path: /config/custom_zha_quirks/zhaquirks/
```
I tried put in zhaquirks all files from this repo.
I tried put only ts0601_dimmer.py. Tried + init.py. Tried tuya/*.
Tried run setup from script. I tried all, but just cant understand and cant find information about this in google. Everything is simple for everyone - "create a folder, upload files and everything will work", but I don't understand what is wrong.
I has HA OS 8.1 on rpi2 with sonoff usb zigbee dongle.
I has simillar error:
```
Logger: homeassistant.config_entries
Source: custom_zha_quirks/zhaquirks/dimmer/ts0601_dimmer.py:13
First occurred: 22:35:18 (1 occurrences)
Last logged: 22:35:18

Error setting up entry Sonoff Zigbee 3.0 USB Dong - /dev/serial/by-id/usb-ITead_Sonoff_Zigbee_3.0_USB_Dongle_Plus_e6900a727729ec11bf936f7840c9ce8d-if00-port0, s/n: e6900a727729ec11bf936f7840c9ce8d - ITead - 10C4:EA60 for zha
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 335, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/usr/src/homeassistant/homeassistant/components/zha/__init__.py", line 99, in async_setup_entry
    setup_quirks(config)
  File "/usr/local/lib/python3.9/site-packages/zhaquirks/__init__.py", line 409, in setup
    importer.find_module(modname).load_module(modname)
  File "<frozen importlib._bootstrap_external>", line 529, in _check_name_wrapper
  File "<frozen importlib._bootstrap_external>", line 1029, in load_module
  File "<frozen importlib._bootstrap_external>", line 854, in load_module
  File "<frozen importlib._bootstrap>", line 274, in _load_module_shim
  File "<frozen importlib._bootstrap>", line 711, in _load
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/config/custom_zha_quirks/zhaquirks/dimmer/ts0601_dimmer.py", line 13, in <module>
    from zhaquirks.tuya import NoManufacturerCluster, TuyaDimmerSwitch
**ImportError: cannot import name 'NoManufacturerCluster' from 'zhaquirks.tuya' (/usr/local/lib/python3.9/site-packages/zhaquirks/tuya/__init__.py)**
```
I understand that he cannot find the library and import them, but how can I install it in the right place? What am I missing?
Please give me guide how i can install this quirks. Thanks a lot ! Please help me

**Comment by dmulcahey** [comment]
Quirks are included by default in HA

**Reply by BashSer** [reply]
I need only write in configuration.yaml:
```
zha:
  custom_quirks_path: /config/custom_zha_quirks/zhaquirks/dimmer/
```
Put in this folder ts0601_dimmer.py and reboot, right ? (+ tried put init.py in this folder, no changes)
Why i has this error ? What is missing ? My head is blow.

![изображение](https://user-images.githubusercontent.com/37932617/171275611-a35e654c-8418-4e17-9847-78bc8fc7871c.png)

**Comment by madspoulsen** [comment]
Yes i'm facing the same issue..
The quirk file is places in the folder /config/custom_zha_quirks/ and is being read, but something is missing..

022-06-07 02:45:19 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Sonoff Zigbee 3.0 USB Dong - /dev/serial/by-id/usb-ITead_Sonoff_Zigbee_3.0_USB_Dongle_Plus_6ad7b52d7112ec11ac2220c7bd930c07-if00-port0, s/n: 6ad7b52d7112ec11ac2220c7bd930c07 - ITead - 10C4:EA60 for zha
File "/usr/src/homeassistant/homeassistant/components/zha/__init__.py", line 99, in async_setup_entry
File "/usr/local/lib/python3.9/site-packages/zhaquirks/__init__.py", line 409, in setup
File "/config/custom_zha_quirks/TS0601_Electric_Heating_0x0051.py", line 37, in <module>

**Reply by MattWestb** [reply]
Something is wrong with this line` TS0601_Electric_Heating_0x0051.py", line 37` or more likely somthing its linking to.
If its need one ` __INIT__ ` you must updating it in the HA container or making changes in your import for it to getting it working OK.

**Reply by madspoulsen** [reply]
Can you explain the purpose of INIT file and its relation to the custom quirks, and how to deal with update its content? I can also see that there is multiple init files.. 🈂️

**Reply by madspoulsen** [reply]
OK, now i spent 2 evenings and it was fun. I got the quirk to load by removing "manufacturer" from the class definition line 37. It has to do with a new implementation from HASS update earlier this year, which caused ZHA crash if a custom quirk mentioned manufacturer_attributes in the beginning..

As this thread has it, the device is lagging functionality to control fan speed and some other. 
If anyone is up for making it with me, let me hear from you.

I can see the fan speed being reported from the device via the ZHA log, but it comes in as "No explicit handler for cluster command 0x01", as expected.
So it should be fairly straightforward to make the handler quirk. I'm just not strong in the syntax (or structure) for quirks.


[0x3223	1	0xef00] Received ZCL frame	 b'\x09\x18\x01\x00\x5E\x1C\x04\x00\x01\x01'	
[0x3223	1	0xef00] Decoded ZCL frame header	 ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.CLUSTER_COMMAND	 1>, is_manufacturer_specific=0, is_reply=1, disable_default_response=0, reserved=0, *is_cluster=True, *is_general=False), tsn=24, command_id=1, *is_reply=True)
[0x3223	1	0xef00] Unknown cluster command 1 b'\x00^\x1c\x04\x00\x01\x01'		
[0x3223	1	0xef00] Received command 0x01 (TSN 24)	 b'\x00^\x1c\x04\x00\x01\x01'	
[0x3223	1	0xef00] No explicit handler for cluster command 0x01	 b'\x00^\x1c\x04\x00\x01\x01'	

Offering a $50 gift card for a solution that provides fan speed, on/off, mode, set-temp, current-temp.
I have the device here and Teamviewer.

**Reply by MattWestb** [reply]
Every level of the python tree is having one  `__INIT__.py` file that is like one configuring file and is having common things that the next level is needing in more quirks so dont need putting the same code in more quirks only linking  the function from the INIT.

If you is running on old HA install and the quirk is of current standard you is missing some function from the INIT that is being added or changed..

You can copy one current version of the file to the HA container with this method https://github.com/zigpy/zha-device-handlers/discussions/693 but i recommending updating (Z)HA to current release for getting all things being at the same level and not getting problem with different version that is not working together.

The INIT file must being in the HA container or its not working if the quirk is not patched for reading the local copy of it that is not standard.

Also deleting all folders and files in `/config/deps/` that can overriding installed HA components and also old local quirk that is merged in ZHA shall being deleted then upgrading to HA with it in so you dont getting prblem in the future.

**Reply by madspoulsen** [reply]
Yea, about that. Can you fix it for me against payment?

**Reply by madspoulsen** [reply]
ok switched to Z2M. Seemed a better fit for me at this time.

**Comment by MattWestb** [comment]
For quirks using `NoManufacturerCluster` need have HA 2022.6 that have implanting it in the tuya INIT https://github.com/zigpy/zha-device-handlers/pull/1572 or must patching it by hand in the HA container for getting it working or patching the import from local tuya INIT.
